src/txn.h

```diff
@@ -31,26 +36,31 @@ typedef enum {
 	TXN_ELEM_TXN_SIG,
 } txnElemType_e;
 
+typedef struct {
+    txnElemType_e elemType; // type of most-recently-seen element
+
+    uint8_t outVal[128];    // most-recently-seen currency value, in decimal
+    uint8_t valLen;         // length of outVal
+    uint8_t outAddr[77];    // most-recently-seen address
+} txn_elem_t;
+
 // txn_state_t is a helper object for computing the SigHash of a streamed
 // transaction.
 typedef struct {
 	uint8_t buf[510]; // holds raw tx bytes; large enough for two 0xFF reads
```

Comment:
if we're no longer streaming, perhaps we could use a `union` here? As in, use the same memory to store both `buf` and `elements`. The unparsed elements in `buf` should always take up more space than their parsed equivalent, so we shouldn't be at risk of overwriting unparsed bytes.

Reply:
But there can be multiple reads.  So if we're the Nano S and we've read in say 10 elements (10*(4 + 24 + 32) = 600 bytes) but we still need to read in more elements then we'd have a problem.  Also I think it might be undefined behavior to use a union like that.  Or maybe I'm misunderstanding something.

Reply:
ah, right